File: server/internal/permissions/permissions.go
package permissions

// Permission bitfield constants.
const (
	KickMembers    int64 = 1 << 0
	BanMembers     int64 = 1 << 1
	ManageRoles    int64 = 1 << 2
	Administrator  int64 = 1 << 3
	ManageEmojis   int64 = 1 << 4
	ManageChannels int64 = 1 << 5
	ManageMessages   int64 = 1 << 6
	TimeoutMembers   int64 = 1 << 7
	ViewAuditLog     int64 = 1 << 8
	ExemptSlowMode   int64 = 1 << 9
	StreamVideo      int64 = 1 << 10
	ManageSoundboard int64 = 1 << 11
	AddReactions     int64 = 1 << 12
	ViewChannel      int64 = 1 << 13
	SendMessages     int64 = 1 << 14
	Connect          int64 = 1 << 15 // voice channel connect
	MentionEveryone  int64 = 1 << 16

	// New permissions (bits 17–28).
	ManageServer       int64 = 1 << 17 // server settings without full admin
	CreateInvite       int64 = 1 << 18
	EmbedLinks         int64 = 1 << 19 // controls link preview generation
	AttachFiles        int64 = 1 << 20
	ReadMessageHistory int64 = 1 << 21
	UseExternalEmojis  int64 = 1 << 22
	Speak              int64 = 1 << 23 // voice: separate from Connect
	MuteMembers        int64 = 1 << 24 // voice moderation
	DeafenMembers      int64 = 1 << 25 // voice moderation
	MoveMembers        int64 = 1 << 26 // voice: move users between channels
	ChangeNickname     int64 = 1 << 27 // self nickname
	ManageNicknames    int64 = 1 << 28 // others' nicknames
	ManageWebhooks     int64 = 1 << 29 // create/edit/delete webhooks

	// AllPermissions is the union of all defined permissions.
	AllPermissions = KickMembers | BanMembers | ManageRoles | Administrator |
		ManageEmojis | ManageChannels | ManageMessages | TimeoutMembers |
		ViewAuditLog | ExemptSlowMode | StreamVideo | ManageSoundboard |
		AddReactions | ViewChannel | SendMessages | Connect | MentionEveryone |
		ManageServer | CreateInvite | EmbedLinks | AttachFiles |
		ReadMessageHistory | UseExternalEmojis | Speak | MuteMembers |
		DeafenMembers | MoveMembers | ChangeNickname | ManageNicknames |
		ManageWebhooks

	// ChannelScopedPermissions contains the bits valid in permission overrides.
	// Server-wide permissions (NOT overridable per-channel):
	//   Administrator, ManageServer, ManageRoles, ManageChannels,
	//   KickMembers, BanMembers, TimeoutMembers, ManageNicknames,
	//   ViewAuditLog, ChangeNickname
	ChannelScopedPermissions = ManageMessages | ExemptSlowMode | StreamVideo |
		ManageSoundboard | AddReactions | ViewChannel | SendMessages |
		Connect | MentionEveryone | ManageEmojis | EmbedLinks | AttachFiles |
		ReadMessageHistory | UseExternalEmojis | Speak | MuteMembers |
		DeafenMembers | MoveMembers | ManageWebhooks

	// DefaultEveryonePermissions is the default permission set for the @everyone role.
	DefaultEveryonePermissions = ViewChannel | SendMessages | Connect | Speak |
		AddReactions | ReadMessageHistory | EmbedLinks | AttachFiles |
		UseExternalEmojis | CreateInvite | ChangeNickname
)

// Intentionally excluded Discord permissions (no corresponding Meza feature):
// - SendTTSMessages: no TTS feature
// - UseApplicationCommands: no bot/application framework
// - ManageEvents: no scheduled events
// - ManageThreads/CreatePublicThreads/CreatePrivateThreads/SendMessagesInThreads: no threads
// - UseVAD: not a meaningful restriction
// - PrioritySpeaker: not planned for launch
// - ViewGuildInsights: no analytics
// - RequestToSpeak: no stage channels

// Has returns true if combined has the given permission.
// Administrator grants all permissions.
func Has(combined, perm int64) bool {
	if combined&Administrator != 0 {
		return true
	}
	return combined&perm != 0
}

// Combine returns the bitwise OR of all permissions.
func Combine(perms ...int64) int64 {
	var c int64
	for _, p := range perms {
		c |= p
	}
	return c
}

// Validate rejects unknown bits to prevent future permission injection.
func Validate(perms int64) bool {
	return perms & ^AllPermissions == 0
}

// ValidateChannelScoped rejects bits that are not valid for permission overrides.
func ValidateChannelScoped(perms int64) bool {
	return perms & ^ChannelScopedPermissions == 0
}

// Source describes where a permission bit comes from.
type Source struct {
	Permission int64
	Granted    bool
	SourceType int // maps to PermissionSourceType proto enum values
	SourceName string
}

// Source type constants matching the proto enum PermissionSourceType.
const (
	SourceBaseRole         = 1
	SourceRoleOverride     = 2
	SourceCategoryOverride = 3
	SourceChannelOverride  = 4
	SourceUserOverride     = 5
	SourceAdministrator    = 6
	SourceOwner            = 7
)

// AttributeSourcesInput extends ResolveInput with name information for source attribution.
type AttributeSourcesInput struct {
	ResolveInput
	EveryoneName string   // usually "@everyone"
	RoleNames    []string // parallel to RolePerms
}

// AttributeSources walks the same resolution logic as ResolveEffective but records
// the source of each permission bit. Call this after ResolveEffective when you need
// to explain WHY each permission is allowed or denied.
func AttributeSources(input AttributeSourcesInput, nowUnix int64) []Source {
	var sources []Source

	// Owner bypass — all permissions come from ownership.
	if input.IsOwner {
		for bit := int64(1); bit <= AllPermissions; bit <<= 1 {
			if AllPermissions&bit != 0 {
				sources = append(sources, Source{Permission: bit, Granted: true, SourceType: SourceOwner, SourceName: "Server Owner"})
			}
		}
		return sources
	}

	// Compute base permissions.
	base := input.EveryonePerms
	for _, rp := range input.RolePerms {
		base |= rp
	}

	// Administrator short-circuit.
	if base&Administrator != 0 {
		for bit := int64(1); bit <= AllPermissions; bit <<= 1 {
			if AllPermissions&bit != 0 {
				sources = append(sources, Source{Permission: bit, Granted: true, SourceType: SourceAdministrator, SourceName: "Administrator"})
			}
		}
		return sources
	}

	// Track which role(s) grant each bit at the base level.
	effective := base

	// For each defined permission bit, record the base source.
	bitSources := make(map[int64]Source)
	for bit := int64(1); bit <= AllPermissions; bit <<= 1 {
		if AllPermissions&bit == 0 {
			continue
		}
		// Check @everyone first.
		if input.EveryonePerms&bit != 0 {
			bitSources[bit] = Source{Permission: bit, Granted: true, SourceType: SourceBaseRole, SourceName: input.EveryoneName}
			continue
		}
		// Check each assigned role.
		granted := false
		for i, rp := range input.RolePerms {
			if rp&bit != 0 {
				name := "Role"
				if i < len(input.RoleNames) {
					name = input.RoleNames[i]
				}
				bitSources[bit] = Source{Permission: bit, Granted: true, SourceType: SourceBaseRole, SourceName: name}
				granted = true
				break
			}
		}
		if !granted {
			bitSources[bit] = Source{Permission: bit, Granted: false, SourceType: SourceBaseRole, SourceName: input.EveryoneName}
		}
	}

	// Timeout strip — if timed out, only ViewChannel + ReadMessageHistory.
	if input.TimedOutUntil > 0 && nowUnix < input.TimedOutUntil {
		for bit := int64(1); bit <= AllPermissions; bit <<= 1 {
			if AllPermissions&bit == 0 {
				continue
			}
			granted := bit == ViewChannel || bit == ReadMessageHistory
			sources = append(sources, Source{Permission: bit, Granted: granted, SourceType: SourceBaseRole, SourceName: "Timed Out"})
		}
		return sources
	}

	// Apply category role overrides.
	for _, ovr := range input.GroupRoleOverrides {
		allow := ovr.Allow & ChannelScopedPermissions
		deny := ovr.Deny & ChannelScopedPermissions
		for bit := int64(1); bit <= ChannelScopedPermissions; bit <<= 1 {
			if deny&bit != 0 {
				effective &= ^bit
				bitSources[bit] = Source{Permission: bit, Granted: false, SourceType: SourceCategoryOverride, SourceName: "Category Override"}
			}
			if allow&bit != 0 {
				effective |= bit
				bitSources[bit] = Source{Permission: bit, Granted: true, SourceType: SourceCategoryOverride, SourceName: "Category Override"}
			}
		}
	}

	// Apply channel role overrides.
	for _, ovr := range input.ChannelRoleOverrides {
		allow := ovr.Allow & ChannelScopedPermissions
		deny := ovr.Deny & ChannelScopedPermissions
		for bit := int64(1); bit <= ChannelScopedPermissions; bit <<= 1 {
			if deny&bit != 0 {
				effective &= ^bit
				bitSources[bit] = Source{Permission: bit, Granted: false, SourceType: SourceChannelOverride, SourceName: "Channel Override"}
			}
			if allow&bit != 0 {
				effective |= bit
				bitSources[bit] = Source{Permission: bit, Granted: true, SourceType: SourceChannelOverride, SourceName: "Channel Override"}
			}
		}
	}

	// Apply category user override.
	if input.GroupUserOverride != nil {
		allow := input.GroupUserOverride.Allow & ChannelScopedPermissions
		deny := input.GroupUserOverride.Deny & ChannelScopedPermissions
		for bit := int64(1); bit <= ChannelScopedPermissions; bit <<= 1 {
			if deny&bit != 0 {
				effective &= ^bit
				bitSources[bit] = Source{Permission: bit, Granted: false, SourceType: SourceUserOverride, SourceName: "Category User Override"}
			}
			if allow&bit != 0 {
				effective |= bit
				bitSources[bit] = Source{Permission: bit, Granted: true, SourceType: SourceUserOverride, SourceName: "Category User Override"}
			}
		}
	}

	// Apply channel user override.
	if input.ChannelUserOverride != nil {
		allow := input.ChannelUserOverride.Allow & ChannelScopedPermissions
		deny := input.ChannelUserOverride.Deny & ChannelScopedPermissions
		for bit := int64(1); bit <= ChannelScopedPermissions; bit <<= 1 {
			if deny&bit != 0 {
				effective &= ^bit
				bitSources[bit] = Source{Permission: bit, Granted: false, SourceType: SourceUserOverride, SourceName: "Channel User Override"}
			}
			if allow&bit != 0 {
				effective |= bit
				bitSources[bit] = Source{Permission: bit, Granted: true, SourceType: SourceUserOverride, SourceName: "Channel User Override"}
			}
		}
	}

	// ViewChannel denied → all denied.
	if effective&ViewChannel == 0 {
		for bit := int64(1); bit <= AllPermissions; bit <<= 1 {
			if AllPermissions&bit == 0 {
				continue
			}
			src := bitSources[bit]
			src.Granted = false
			sources = append(sources, src)
		}
		return sources
	}

	// Collect final sources.
	for bit := int64(1); bit <= AllPermissions; bit <<= 1 {
		if AllPermissions&bit == 0 {
			continue
		}
		if src, ok := bitSources[bit]; ok {
			sources = append(sources, src)
		}
	}

	return sources
}

// Override represents an allow/deny permission pair for a channel or category override.
type Override struct {
	Allow int64
	Deny  int64
}

// ResolveInput contains all data needed to compute effective permissions.
type ResolveInput struct {
	EveryonePerms int64   // @everyone role permissions
	RolePerms     []int64 // permissions from each assigned role
	IsOwner       bool    // server owner bypass
	TimedOutUntil int64   // unix timestamp; 0 means not timed out

	// Channel-level overrides (leave empty for server-level checks).
	GroupRoleOverrides   []Override // category overrides for member's roles
	ChannelRoleOverrides []Override // channel overrides for member's roles
	GroupUserOverride    *Override  // category override for this specific user
	ChannelUserOverride  *Override  // channel override for this specific user
}

// ResolveEffective computes effective permissions following Discord's resolution algorithm:
//  1. Owner bypass → AllPermissions
//  2. Base = @everyone | all role perms
//  3. Administrator short-circuit → AllPermissions
//  4. Timeout strip → ViewChannel | ReadMessageHistory only
//  5. Apply category role overrides (aggregated)
//  6. Apply channel role overrides (aggregated, more specific wins)
//  7. Apply category user override
//  8. Apply channel user override (most specific, wins over everything)
//  9. ViewChannel denied → 0
func ResolveEffective(input ResolveInput, nowUnix int64) int64 {
	// 1. Owner bypass.
	if input.IsOwner {
		return AllPermissions
	}

	// 2. Base = @everyone | all role perms.
	base := input.EveryonePerms
	for _, rp := range input.RolePerms {
		base |= rp
	}

	// 3. Administrator short-circuit.
	if base&Administrator != 0 {
		return AllPermissions
	}

	// 4. Timeout strip — timed-out users get only ViewChannel + ReadMessageHistory.
	if input.TimedOutUntil > 0 && nowUnix < input.TimedOutUntil {
		return ViewChannel | ReadMessageHistory
	}

	// 5. Apply category role overrides (aggregate all roles).
	var groupAllow, groupDeny int64
	for _, ovr := range input.GroupRoleOverrides {
		groupAllow |= (ovr.Allow & ChannelScopedPermissions)
		groupDeny |= (ovr.Deny & ChannelScopedPermissions)
	}
	base = (base & ^groupDeny) | groupAllow

	// 6. Apply channel role overrides.
	var chAllow, chDeny int64
	for _, ovr := range input.ChannelRoleOverrides {
		chAllow |= (ovr.Allow & ChannelScopedPermissions)
		chDeny |= (ovr.Deny & ChannelScopedPermissions)
	}
	base = (base & ^chDeny) | chAllow

	// 7. Apply category user override (if any).
	if input.GroupUserOverride != nil {
		allow := input.GroupUserOverride.Allow & ChannelScopedPermissions
		deny := input.GroupUserOverride.Deny & ChannelScopedPermissions
		base = (base & ^deny) | allow
	}

	// 8. Apply channel user override (most specific, wins over everything).
	if input.ChannelUserOverride != nil {
		allow := input.ChannelUserOverride.Allow & ChannelScopedPermissions
		deny := input.ChannelUserOverride.Deny & ChannelScopedPermissions
		base = (base & ^deny) | allow
	}

	// 9. ViewChannel denied → no permissions in this channel.
	if base&ViewChannel == 0 {
		return 0
	}

	return base
}

File: client/packages/core/src/store/permissions.ts
export const Permissions = {
  KICK_MEMBERS: 1n << 0n,
  BAN_MEMBERS: 1n << 1n,
  MANAGE_ROLES: 1n << 2n,
  ADMINISTRATOR: 1n << 3n,
  MANAGE_EMOJIS: 1n << 4n,
  MANAGE_CHANNELS: 1n << 5n,
  MANAGE_MESSAGES: 1n << 6n,
  TIMEOUT_MEMBERS: 1n << 7n,
  VIEW_AUDIT_LOG: 1n << 8n,
  EXEMPT_SLOW_MODE: 1n << 9n,
  STREAM_VIDEO: 1n << 10n,
  MANAGE_SOUNDBOARD: 1n << 11n,
  ADD_REACTIONS: 1n << 12n,
  VIEW_CHANNEL: 1n << 13n,
  SEND_MESSAGES: 1n << 14n,
  CONNECT: 1n << 15n,
  MENTION_EVERYONE: 1n << 16n,
  MANAGE_SERVER: 1n << 17n,
  CREATE_INVITE: 1n << 18n,
  EMBED_LINKS: 1n << 19n,
  ATTACH_FILES: 1n << 20n,
  READ_MESSAGE_HISTORY: 1n << 21n,
  USE_EXTERNAL_EMOJIS: 1n << 22n,
  SPEAK: 1n << 23n,
  MUTE_MEMBERS: 1n << 24n,
  DEAFEN_MEMBERS: 1n << 25n,
  MOVE_MEMBERS: 1n << 26n,
  CHANGE_NICKNAME: 1n << 27n,
  MANAGE_NICKNAMES: 1n << 28n,
  MANAGE_WEBHOOKS: 1n << 29n,
} as const;

/** OR of all 30 permission bits. */
export const ALL_PERMISSIONS = Object.values(Permissions).reduce(
  (acc, v) => acc | v,
  0n,
);

export const CHANNEL_SCOPED_PERMISSIONS =
  Permissions.MANAGE_MESSAGES |
  Permissions.EXEMPT_SLOW_MODE |
  Permissions.STREAM_VIDEO |
  Permissions.MANAGE_SOUNDBOARD |
  Permissions.ADD_REACTIONS |
  Permissions.VIEW_CHANNEL |
  Permissions.SEND_MESSAGES |
  Permissions.CONNECT |
  Permissions.MENTION_EVERYONE |
  Permissions.MANAGE_EMOJIS |
  Permissions.EMBED_LINKS |
  Permissions.ATTACH_FILES |
  Permissions.READ_MESSAGE_HISTORY |
  Permissions.USE_EXTERNAL_EMOJIS |
  Permissions.SPEAK |
  Permissions.MUTE_MEMBERS |
  Permissions.DEAFEN_MEMBERS |
  Permissions.MOVE_MEMBERS |
  Permissions.MANAGE_WEBHOOKS;

export function hasPermission(combined: bigint, perm: bigint): boolean {
  if ((combined & Permissions.ADMINISTRATOR) !== 0n) return true;
  return (combined & perm) !== 0n;
}

export function validateChannelScoped(perms: bigint): boolean {
  return (perms & ~CHANNEL_SCOPED_PERMISSIONS) === 0n;
}

export function validatePermissions(perms: bigint): boolean {
  return (perms & ~ALL_PERMISSIONS) === 0n;
}

export type PermCategory =
  | 'general'
  | 'text'
  | 'voice'
  | 'moderation'
  | 'server';

/** Derived union type for type-safe permission keys. */
export type PermissionKey = keyof typeof PERMISSION_INFO;

/** Channel type → which permission categories to show in override editor. */
export const CHANNEL_TYPE_CATEGORIES: Record<'text' | 'voice', PermCategory[]> =
  {
    text: ['general', 'text'],
    voice: ['general', 'voice'],
  };

/** Display metadata for each permission category. */
export const CATEGORY_META: Record<
  PermCategory,
  { label: string; icon: string }
> = {
  general: { label: 'General Permissions', icon: 'globe' },
  text: { label: 'Text Channel Permissions', icon: 'message-square' },
  voice: { label: 'Voice Channel Permissions', icon: 'volume-2' },
  moderation: { label: 'Moderation Permissions', icon: 'shield' },
  server: { label: 'Server Management', icon: 'settings' },
};

interface PermInfo {
  name: string;
  description: string;
  category: PermCategory;
}

export const PERMISSION_INFO = {
  KICK_MEMBERS: {
    name: 'Kick Members',
    description: 'Allows kicking members from the server',
    category: 'moderation',
  },
  BAN_MEMBERS: {
    name: 'Ban Members',
    description: 'Allows banning members from the server',
    category: 'moderation',
  },
  MANAGE_ROLES: {
    name: 'Manage Roles',
    description: 'Allows creating, editing, and deleting roles',
    category: 'server',
  },
  ADMINISTRATOR: {
    name: 'Administrator',
    description: 'Grants all permissions and bypasses channel overrides',
    category: 'server',
  },
  MANAGE_EMOJIS: {
    name: 'Manage Emojis',
    description: 'Allows managing custom emojis',
    category: 'server',
  },
  MANAGE_CHANNELS: {
    name: 'Manage Channels',
    description: 'Allows creating, editing, and deleting channels',
    category: 'server',
  },
  MANAGE_MESSAGES: {
    name: 'Manage Messages',
    description: "Allows deleting other members' messages and pinning messages",
    category: 'text',
  },
  TIMEOUT_MEMBERS: {
    name: 'Timeout Members',
    description: 'Allows temporarily restricting member permissions',
    category: 'moderation',
  },
  VIEW_AUDIT_LOG: {
    name: 'View Audit Log',
    description: 'Allows viewing the server audit log',
    category: 'server',
  },
  EXEMPT_SLOW_MODE: {
    name: 'Exempt from Slow Mode',
    description: 'Allows sending messages without slow mode restrictions',
    category: 'text',
  },
  STREAM_VIDEO: {
    name: 'Stream Video',
    description: 'Allows sharing screen in voice channels',
    category: 'voice',
  },
  MANAGE_SOUNDBOARD: {
    name: 'Manage Soundboard',
    description: 'Allows managing server soundboard sounds',
    category: 'voice',
  },
  ADD_REACTIONS: {
    name: 'Add Reactions',
    description: 'Allows adding emoji reactions to messages',
    category: 'text',
  },
  VIEW_CHANNEL: {
    name: 'View Channel',
    description: 'Allows viewing a channel',
    category: 'general',
  },
  SEND_MESSAGES: {
    name: 'Send Messages',
    description: 'Allows sending messages in text channels',
    category: 'text',
  },
  CONNECT: {
    name: 'Connect',
    description: 'Allows joining voice channels',
    category: 'voice',
  },
  MENTION_EVERYONE: {
    name: 'Mention Everyone',
    description: 'Allows using @everyone mentions',
    category: 'text',
  },
  MANAGE_SERVER: {
    name: 'Manage Server',
    description: 'Allows editing server name, icon, and settings',
    category: 'server',
  },
  CREATE_INVITE: {
    name: 'Create Invite',
    description: 'Allows creating invite links',
    category: 'general',
  },
  EMBED_LINKS: {
    name: 'Embed Links',
    description: 'Allows link previews to be displayed',
    category: 'text',
  },
  ATTACH_FILES: {
    name: 'Attach Files',
    description: 'Allows uploading files and images',
    category: 'text',
  },
  READ_MESSAGE_HISTORY: {
    name: 'Read Message History',
    description: 'Allows reading message history in channels',
    category: 'text',
  },
  USE_EXTERNAL_EMOJIS: {
    name: 'Use External Emojis',
    description: 'Allows using emojis from other servers',
    category: 'text',
  },
  SPEAK: {
    name: 'Speak',
    description: 'Allows speaking in voice channels',
    category: 'voice',
  },
  MUTE_MEMBERS: {
    name: 'Mute Members',
    description: 'Allows muting other members in voice channels',
    category: 'voice',
  },
  DEAFEN_MEMBERS: {
    name: 'Deafen Members',
    description: 'Allows deafening other members in voice channels',
    category: 'voice',
  },
  MOVE_MEMBERS: {
    name: 'Move Members',
    description: 'Allows moving members between voice channels',
    category: 'voice',
  },
  CHANGE_NICKNAME: {
    name: 'Change Nickname',
    description: 'Allows changing your own nickname',
    category: 'general',
  },
  MANAGE_NICKNAMES: {
    name: 'Manage Nicknames',
    description: "Allows changing other members' nicknames",
    category: 'moderation',
  },
  MANAGE_WEBHOOKS: {
    name: 'Manage Webhooks',
    description: 'Allows creating, editing, and deleting webhooks',
    category: 'server',
  },
} as const satisfies Record<string, PermInfo>;

/** Group permission keys by category for UI rendering. */
export const PERMISSIONS_BY_CATEGORY: Record<PermCategory, PermissionKey[]> = {
  general: ['VIEW_CHANNEL', 'CREATE_INVITE', 'CHANGE_NICKNAME'],
  text: [
    'SEND_MESSAGES',
    'EMBED_LINKS',
    'ATTACH_FILES',
    'ADD_REACTIONS',
    'MENTION_EVERYONE',
    'MANAGE_MESSAGES',
    'READ_MESSAGE_HISTORY',
    'USE_EXTERNAL_EMOJIS',
    'EXEMPT_SLOW_MODE',
  ],
  voice: [
    'CONNECT',
    'SPEAK',
    'STREAM_VIDEO',
    'MUTE_MEMBERS',
    'DEAFEN_MEMBERS',
    'MOVE_MEMBERS',
    'MANAGE_SOUNDBOARD',
  ],
  moderation: [
    'KICK_MEMBERS',
    'BAN_MEMBERS',
    'TIMEOUT_MEMBERS',
    'MANAGE_NICKNAMES',
  ],
  server: [
    'ADMINISTRATOR',
    'MANAGE_SERVER',
    'MANAGE_CHANNELS',
    'MANAGE_ROLES',
    'MANAGE_EMOJIS',
    'MANAGE_WEBHOOKS',
    'VIEW_AUDIT_LOG',
  ],
};
